modules/blockbuilder/tenant_store.go

```diff
@@ -157,18 +157,32 @@ func (s *tenantStore) Flush(ctx context.Context, store tempodb.Writer) error {
 	return nil
 }
 
+// Adjust the time range based on when the record was added to the partition, factoring in slack and cycle duration.
+// Any span with a start or end time outside this range will be constrained to the valid limits.
 func (s *tenantStore) adjustTimeRangeForSlack(start, end time.Time) (time.Time, time.Time) {
 	startOfRange := s.startTime.Add(-s.slackDuration)
 	endOfRange := s.startTime.Add(s.slackDuration + s.cycleDuration)
 
 	warn := false
 	if start.Before(startOfRange) {
 		warn = true
-		start = s.startTime
+		start = startOfRange
 	}
-	if end.After(endOfRange) || end.Before(start) {
+	// clock skew, missconfiguration or simply data tampering
+	if start.After(endOfRange) {
 		warn = true
-		end = s.startTime
+		start = endOfRange
+	}
+	// this can happen with old spans added to Tempo
+	// setting it to start to not jump forward unexpectedly
+	if end.Before(start) {
+		warn = true
+		end = start
+	}
```

Comment:
This is just wrong data, no? I don't think we have to fix it, or at least it's not related to ingestion slack.

Reply:
Did you mean `end.Before(startOfRange)` instead?

```suggestion
	if end.Before(startOfRange) {
		warn = true
		end = start
	}
```

Reply:
Not wrong data, old data. Imagine someone wants to add old spans into Tempo
I did a very similar fix a time ago:
https://github.com/grafana/tempo/pull/3954
We need to correct the end date or the block will be wrong

It's checked against the ``start`` which can be already corrected before. That way we ensure that the start is never greater than the end. If we check it against ``startOfRange`` we may miss an edge case where startDate < end < start

Reply:
using `start` instead of `startOfRange` makes sense to me because it addresses both scenarios. If the `start` is before the `startOfRange` it has been already adjusted. In the edge case where `end < start < startOfRange` it fixes it as well

Reply:
>In the edge case where end < startOfRange < start it fixes it as well

This is what I meant by wrong data, `end < start` is logically incorrect.

Reply:
Internally during a test block builders flushed several blocks in a row with meta.start==meta.end.  I'm not sure that this change is the fix, because the only related change is the case `end after endOfRange`. Previously it was set to `start` and now it's `endOfRange`.  For this to be the issue it would have to mean that every trace in the block was a long-running one that exceeds the buffer `s.startTime.Add(s.slackDuration + s.cycleDuration)`.  I'd like to dig more and see if there is another explanation.